src/email/mod.rs

```diff
@@ -575,13 +605,61 @@ impl EmailBuilder {
             assert!(self.sender_header.is_some());
         }
         // Add the sender header, if any.
-        if let Some(v) = self.sender_header {
+        if let Some(ref v) = self.sender_header {
             self.message.add_header(("Sender", v.to_string().as_ref()));
         }
+        // Calculate the envelope
+        let envelope = match self.envelope {
+            Some(e) => e,
+            None => {
+                // we need to generate the envelope
+                let mut e = Envelope::new();
+                // add all receivers in to_header and cc_header
+                for a in self.to_header.iter().chain(self.cc_header.iter()) {
+                    match *a {
+                        Address::Mailbox(ref m) => e.add_to(m.address.clone()),
+                        Address::Group(_, ref ms) => {
+                            for m in ms.iter() {
+                                e.add_to(m.address.clone());
+                            }
+                        }
+                    }
+                }
+                if e.to.is_empty() {
+                    return Err(Error::MissingTo);
+                }
+                e.set_from(match self.sender_header {
+                    Some(x) => x.address.clone(), // if we have a sender_header, use it
+                    None => {
+                        debug_assert!(self.from_header.len()<=1); // else we'd have sender_header
+                        match self.from_header.first() {
+                            Some(a) => match *a {
+                                // if we have a sender_header
+                                Address::Mailbox(ref m) => m.address.clone(), // use it
+                                Address::Group(_,ref ms) => match ms.first() {
+                                    // if it's an author group, use the first author
+                                    Some(m) => m.address.clone(),
+                                    // for an empty author group (the rarest of the rare cases)
+                                    None => String::new(), // empty envelope sender
```

Comment:
If we do not implement the non-compliance flag now, I would prefer keeping the error.